neqo-common/src/codec.rs

```diff
@@ -197,12 +202,120 @@ impl<'b> PartialEq<Decoder<'b>> for Decoder<'_> {
 }
 
 /// Encoder is good for building data structures.
-#[derive(Clone, Default, PartialEq, Eq)]
-pub struct Encoder {
-    buf: Vec<u8>,
+#[derive(Clone, PartialEq, Eq)]
+pub struct Encoder<B = Vec<u8>> {
+    buf: B,
+}
+
+impl Default for Encoder {
+    fn default() -> Self {
+        Self { buf: Vec::new() }
+    }
+}
+
+impl<B: std::io::Write> Encoder<B> {
```

Comment:
Rather than rely on `io::Write`, can you instead use `AsMut<[u8]>` as the target type of `B`?  That would remove all of the concerns about writing and restore all the previous concerns about indexing.  But it would make some of these functions easier to copy over.  You'd need to track the write position, so it's a smidge more memory for the `Vec` case, but I think that's acceptable on balance.

Reply:
I chose `io::Write` instead of `AsMut<[u8]>` for the `Encoder<Vec<u8>>` case, where the `Vec<u8>` is empty at first, e.g. due to the caller not knowing how much memory is needed. Note that this is the most common case across all of Neqo today.

**[`Vec<u8>`'s `io::Write`](https://doc.rust-lang.org/src/std/io/impls.rs.html#503-506) implementation will grow the `Vec`. `AsMut<[u8]>` can not.**

`encode_vvec_with` implementation is the only function that really changes in this PR. I still need to do more benchmarks. Intuitively it shouldn't come with a performance hit. Where ever it does, we can simply replace `encode_vvec_with`.

With the above in mind, can you think of a better way forward @martinthomson?

Reply:
Friendly ping @martinthomson.

Reply:
Thanks for the input. I updated the pull request to use a trait and multiple implementations.